ironfish/src/network/peers/addressManager.ts

```diff
@@ -12,71 +13,114 @@ import { PeerManager } from './peerManager'
  * and provides functionality for persistence of said data.
  */
 export class AddressManager {
+  LIMIT = 50
+
   hostsStore: HostsStore
   peerManager: PeerManager
+  peerIdentityMap: Map<Identity, PeerAddress>
 
   constructor(hostsStore: HostsStore, peerManager: PeerManager) {
     this.hostsStore = hostsStore
     this.peerManager = peerManager
-  }
+    // load prior peers from disk
+    this.peerIdentityMap = new Map<string, PeerAddress>()
 
-  get priorConnectedPeerAddresses(): ReadonlyArray<Readonly<PeerAddress>> {
-    return this.hostsStore.getArray('priorPeers')
-  }
+    let priorPeers = this.hostsStore.getArray('priorPeers').filter((peer) => {
+      if (peer.identity === null || peer.address === null || peer.port === null) {
+        return false
+      }
+      // Backwards compatible change: if lastAddedTimestamp is undefined or null,
+      // set it to the current time.
+      if (peer.lastAddedTimestamp === undefined) {
+        peer.lastAddedTimestamp = 0
+      }
+      return true
+    })
 
-  /**
-   * Returns a peer address for a disconnected peer by using current peers to
-   * filter out peer addresses. It attempts to find a previously-connected
-   * peer address that is not part of an active connection.
-   */
-  getRandomDisconnectedPeerAddress(peerIdentities: string[]): PeerAddress | null {
-    if (this.priorConnectedPeerAddresses.length === 0) {
-      return null
+    // If there are more than 50 peers, we remove
+    // extra peers from the list. This should only happen during
+    // the first time the node is started after this change is implemented.
+    if (priorPeers.length > this.LIMIT) {
+      priorPeers = priorPeers.slice(0, this.LIMIT)
     }
 
-    const currentPeerIdentities = new Set(peerIdentities)
-
-    const disconnectedPriorAddresses = this.priorConnectedPeerAddresses.filter(
-      (address) => address.identity !== null && !currentPeerIdentities.has(address.identity),
-    )
-    if (disconnectedPriorAddresses.length) {
-      return ArrayUtils.sampleOrThrow(disconnectedPriorAddresses)
+    for (const peer of priorPeers) {
+      this.peerIdentityMap.set(peer.identity, peer)
     }
 
-    return null
+    void this.save()
```

Comment:
Maybe this is worth discussing more, but my first reaction is that these (this one and the others in this file) should not be void, but instead `return this.save()`. Places that explicitly don't want to wait for the promise can use `void` at their call-site, but in most cases we probably _do_ want to wait for the promise.

Reply:
I'm implementing this on my local and trying to think about the changes, will update what I find

Reply:
I do feel like it is more explicit that these functions have been voided. So I made the change.

Reply:
This is in the constructor though so we can't have async code. I think maybe makes sense just to not call `save()` at all in the constructor and wait for it to be saved on the original schedule

Reply:
Although, I think TS doesn't allow constructors to be async. I feel like it's not super critical to add an await here because the addPeer will be called pretty quickly after this instantiates. So, I think it's fine to leave it as void here. 

The change would be to create an `init` function and call that where the address manager is instantiated. I'm OK doing that and it fits the pattern of our other classes

Reply:
@danield9tqh you're right, i'll remove it from the constructor. It's going to be called right after with addPeer